### ISSUE TITLE
Display Issue (black history graph area)

### ISSUE THREAD

**Issue by lguilhot** [gitextensions/gitextensions#6268]
The whole history area is being displayed in black after one second

## Current behaviour

Once opening a new repository or upon checkout of a new branch, the history graph is displayed correctly during about 1 or two seconds
After that delay the whole graph is displayed in black

## Expected behaviour
The graph remains displayed correctly

## Steps to reproduce

Just trying to browse a reository.

Hardware is a Thinkpad Laptop

## Screenshots

![image](https://user-images.githubusercontent.com/47741269/52968095-7f05f880-33ac-11e9-99bd-090d9c4ec888.png)

## Did this work in previous version of GitExtensions

Yes, it was working before. I don't remember the previous version, but there was still not the "branch browsing windows"

## Environment

- Git Extensions 3.0.2.5232
- Build 23b6f51905006ccdde8701591df706284b4155dc
- Git 2.20.1.windows.1
- Microsoft Windows NT 10.0.16299.0
- .NET Framework 4.7.3163.0
- DPI 144dpi (150% scaling)

**Comment by vbjay** [comment]
Do you have submodules?

**Comment by lguilhot** [comment]
Yes in some repositories

**Comment by vbjay** [comment]
Is the .git item in a submodule folder a folder or a file?

**Comment by lguilhot** [comment]
In the given example, I was browsing a submodule folder having a .git item

**Comment by lguilhot** [comment]
Inbetween I opened other submodules that seemed to work properly, but the display became also suddently black as I clicked on an old revision in the graph.

**Comment by vbjay** [comment]
You failed to answer my question. Is .git a folder or a file?

**Comment by lguilhot** [comment]
Ah ok, .git is a folder

**Comment by vbjay** [comment]
That is your problem. You have configured a submodule but cloned a repo instead of using git submodule update --init

**Comment by lguilhot** [comment]
That's right, the submodules were existing, and I decided to gather them into a parent module later.
Is there a way to fix that keeping the history?

**Comment by vbjay** [comment]
Push all branches to the remote. Delete the whole contents of the submodule
folder not the actual submodule folder. In that folder run git submodule
update --init.

Do not ever clone into a submodule.

**Comment by lguilhot** [comment]
Ok, I'll give it a try, thank you for your help.

**Comment by RussKie** [comment]
This seems to be a recurring issue lately. Anything we can do to notify the user about the problem?
/cc: @gitextensions/git-extensions-source

**Comment by vbjay** [comment]
Yes. If .git in a registered submodule is a folder alert that the repos ate
not configured correctly. Point them to a wiki page.

**Comment by RussKie** [comment]
@gerhardol / @amaiorano do you think it is easy to do?

**Comment by amaiorano** [comment]
From what I've understood so far, this happens when there's an extra .git folder anywhere under the tree, right? Probably as an attempt to add a submodule incorrectly, right? I'd like to know how this situation happens. A user would have to explicitly clone another repo in a subfolder, presumably outside of GE? Perhaps we'd need to look recursively for .git folders in the entire tree upon opening a module.

**Comment by gerhardol** [comment]
This is probably more common due to submodules are more common. If it can be reproduced, it should be some way to inform the user at least

**Comment by vbjay** [comment]
I think just a .git folder without submodules is weird but valid.  I think
it is where you have a submodule path and .git is not a file.

**Comment by vbjay** [comment]
![image](https://user-images.githubusercontent.com/3629489/52977160-39d8cb00-339a-11e9-854a-917868bec654.png)

**Comment by amaiorano** [comment]
> I think just a .git folder without submodules is weird but valid. I think it is where you have a submodule path and .git is not a file.

You're probably right. Do we know _why_ this results in a black screen? Are we throwing an exception that we handle silently?

Again, I have to wonder how this happens. A user would explicitly git clone a submodule in the location that the submodule should be in?

**Comment by vbjay** [comment]
Think.  It can't find the git objects.  YOU DO NOT CLONE in a submodule.  You run git submodule update --init.  Which clones in the parent repo .git folder in the modules folder.  Look at the git extensions repo.  Look inside the modules folder in .git.  

If you tell git that there is a git module and it tries to read a folder as a file....  See the problem?

**Comment by RussKie** [comment]
As a hypothesis, we cache rendered revision graph listview items' bitmaps,
and there must be an exception or something else that blocks rendering...

**Comment by amaiorano** [comment]
> Think. It can't find the git objects. YOU DO NOT CLONE in a submodule. You run git submodule update --init. Which clones in the parent repo .git folder in the modules folder. Look at the git extensions repo. Look inside the modules folder in .git.
> 
> If you tell git that there is a git module and it tries to read a folder as a file.... See the problem?

Yes, @vbjay, I understand everything you've said so far :) What I keep asking is, how come this happens? What are user doings to make this occur? And second, if we are trying to read this file that is now a folder, why are we silently failing and showing a black screen. The second question, we can answer easily enough by debugging. The first... I have no idea :)

**Comment by vbjay** [comment]
This is what they are doing

INVALID CONFIG:
```bash
cd parentRepoFolder
git clone ... which creates SomeSubRepo Folder
git submodule add pointing to SomeSubRepo
```

**Comment by amaiorano** [comment]
> This is what they are doing
> 
> INVALID CONFIG:
> 
> ```shell
> cd parentRepoFolder
> git clone ... which creates SomeSubRepo Folder
> git submodule add pointing to SomeSubRepo
> ```

Seems plausible. @lguilhot is this your case?

**Comment by vbjay** [comment]
The other way is they are calling git submodule add, and not seeing their files, so they cd into the submodule folder and git clone.

**Comment by amaiorano** [comment]
Feels like some of these common user errors could be handled by Git proper. Like if you try to clone a repo that is in gitmodules, it could fail and suggest that the user use "git submodule update --init". Or if the user tries to "git submodule add" a repo that has a .git folder, it could fail there as well.

**Comment by vbjay** [comment]
You would think....  :man_shrugging:

**Comment by vbjay** [comment]
Before anyone touches git, they should read https://www.vogella.com/tutorials/Git/article.html

**Comment by vbjay** [comment]
Oh well.  It used to be a full and complete tutorial.  It had contents and everything.  Get the book.

**Comment by gerhardol** [comment]
There should be an error status from Git at one point that should be possible to use to find that something is is wrong with the repo. Someone with time to spare, please dig in.

**Comment by lguilhot** [comment]
@amaiorano 
Hi,
yes, I think this is my situation:
I used to work with existing repositories with some history.
I decided to setup a top project in the parent folder and used (as far as I remember) the command "git submodule add" for adding the existing submodules.
(Note that I'm here in a special case where the remote is not a git server but just a remote network directory with a bare project).

I tried to delete the submodule and run "git submodule update --init", but I got back my .git folder which is probably normal as it exists in the remote location and the problem persists.

**Comment by vbjay** [comment]
It doesn't matter if the submodule points to a folder outside of the parent repo or a GitHub repo. 

Parent Repo
|__folder
|__Submodule folder created by git submodule add pointing to X


X

**Comment by amaiorano** [comment]
> @amaiorano
> Hi,
> yes, I think this is my situation:
> I used to work with existing repositories with some history.
> I decided to setup a top project in the parent folder and used (as far as I remember) the command "git submodule add" for adding the existing submodules.
> (Note that I'm here in a special case where the remote is not a git server but just a remote network directory with a bare project).
> 
> I tried to delete the submodule and run "git submodule update --init", but I got back my .git folder which is probably normal as it exists in the remote location and the problem persists.

Thanks for explaining, @lguilhot. So I think that we can't be expected to handle detecting misuse of Git like this -- this should be done by Git proper. However, we can tackle the black screen thing: figure out what exception is being thrown, and do something better, like warn the user that something's wrong with a dialog, etc.

**Comment by lguilhot** [comment]
Hi,
so i tried to re-create from scratch a parent repository, adding the submodules pointing to the existing remote child repositories.

Now, I have in each submmodule folder a .git file instead of a .git folder.

Nevertheless this didn't solve the black window issue.

**Comment by amaiorano** [comment]
> Hi,
> so i tried to re-create from scratch a parent repository, adding the submodules pointing to the existing remote child repositories.
> 
> Now, I have in each submmodule folder a .git file instead of a .git folder.
> 
> Nevertheless this didn't solve the black window issue.

Just to be clear, this isn't the way you're supposed to add submodules. The usual approach is to start from a clean repo, then use "git submodule add <repo url> <path>" (see [documentation](https://git-scm.com/docs/git-submodule)). Then you can init and update to pull in those submodules to the configured local paths. You're not supposed to start from existing files, as you're trying to do. Again, it would be nice if Git helped here, but I suppose it doesn't, and you end up in a weird state.

**Comment by lguilhot** [comment]
Yes, i think this is what I did, unless I missed something:
- I created a new parent repostitory
- then, in this repository I added submodules , using:
"git submodule add [path to remote bare  repository] [path to submodule]
"git submodule init"

**Comment by RussKie** [comment]
Fixed in #6318
